File: L2AmazonMedium/LongestPalindromeSubstring.java

public class LongestPalindromeSubstring {

    public static String longestPalindrome(String s) {
        if (s == null || s.length() < 1) {
            return "";
        }

        int start = 0, end = 0;

        for (int i = 0; i < s.length(); i++) {
            // Expand around the center for odd-length palindromes
            int len1 = expandAroundCenter(s, i, i);
            // Expand around the center for even-length palindromes
            int len2 = expandAroundCenter(s, i, i + 1);

            // Find the maximum length from both cases
            int len = Math.max(len1, len2);

            // Update the start and end pointers if a longer palindrome is found
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }

        // Return the longest palindromic substring
        return s.substring(start, end + 1);
    }

    private static int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        // Return the length of the palindrome
        return right - left - 1;
    }

    public static void main(String[] args) {
        String s = "babad";
        String result = longestPalindrome(s);
        System.out.println("Longest Palindromic Substring: " + result);
    }
}

File: L2AmazonMedium/output/LongestPalindromeSubstring.cpp
#include <iostream>
#include <string>
using namespace std;

void longestPalSubstr(const string &s)
{
    int start = 0, end = 0;

    for (int i = 0; i < s.length(); i++)
    {
        // Odd-length palindrome
        int l = i, r = i;
        while (l >= 0 && r < s.length() && s[l] == s[r])
        {
            if (r - l + 1 > end - start)
            {
                start = l;
                end = r + 1;
            }
            l--;
            r++;
        }

        // Even-length palindrome
        l = i;
        r = i + 1;
        while (l >= 0 && r < s.length() && s[l] == s[r])
        {
            if (r - l + 1 > end - start)
            {
                start = l;
                end = r + 1;
            }
            l--;
            r++;
        }
    }

    // Print the longest palindromic substring
    cout << "Longest Palindromic Substring: " << s.substr(start, end - start) << endl;
}

int main()
{
    string s = "babad";
    longestPalSubstr(s);
    return 0;
}
